### ISSUE TITLE
Multi-Peak Event Builder should be separated into two classes

### ISSUE THREAD

**Issue by evzayas** [project8/katydid#77]
The multi-peak event builder consists of two steps: (a) group parallel tracks together and (b) combine these groups into an event sequence using head-tail comparison. These functions should each belong to an individual processor, so that the intermediate data object (a group of parallel tracks) is more accessible, and to make easier the integration of secondary track finding analysis in the future. The code is already separated by these two functions:

```
if (! FindMultiPeakTracks())
{
    KTERROR(tclog, "An error occurred while identifying multipeak tracks");
    return false;
}

if (! FindEvents())
{
    KTERROR(tclog, "An error occurred while identifying events");
    return false;
}

KTDEBUG(tclog, "Event building complete");
fEventsDoneSignal();

return true;
```

So the transition should be straight-forward.

**Comment by nsoblath** [comment]
I think this makes sense to do.  You'll need a data object to represent a set of parallel tracks, of course.  You can even include in that class a flag to indicate if it's the first grouping in an event, and that can be modified in the FindEvents stage.  It would need to have at least three values: yes, no, and don't-know-yet.

**Comment by guiguem** [comment]
The group of peaks given to (b) should be an object equivalent to one single track with one single start time, start frequency, end time and end frequency. Between (a) and (b), one could make a simple determination of the nature of the peaks based on the power/width.. and then use this nature to infer on the parameters of the main peak (especially if the main peak is missing). Each group of tracks could then have parameters such as the number of sidebands detected, such as enduser analysis can use this information for more complex analysis...

**Comment by evzayas** [comment]
@Noah- I think that's a good idea as well. This would be a useful flag to cut on in katydid directly, to reduce the complexity of post-processing as Ben said.

@Mathieu- that is the idea I had in mind. Better yet, (b) could have two slots: one for the direct output of (a), and another for the object equivalent to a single track as you describe.

**Comment by guiguem** [comment]
why two slots? not sure to see the point here. the head-tail algorithm is only associating a start time with a end time and combining the associated objects together... not sure what extra informations you could use that you didn't use already in the intermediary step.
My point is that the intermediary step between (a) and (b) should be mandatory, even in a very very basic implementation (based on the power of the track for example). I guess you could keep the possibility of analyzing the direct output of (a), but that should be depreciated since it is not optimal...

**Comment by evzayas** [comment]
I was thinking mostly to maintain the current functionality of the MPEB in full. The second half of it probably will be depreciated once we can develop a good intermediate step, but I see no reason to get rid of it in a hurry.

**Comment by nsoblath** [comment]
The first step should be to replicate the current functionality with two new processors, as Evan suggests.  

We should also consider adding a base class for data objects that are track-like, similar to how we have a base class for objects that are spectrum-like.  This would allow us to simplify the code in processors that act on different track-like objects (single tracks, multi-peak tracks, future objects that add more information, etc).

These two things could be done in parallel.

**Comment by evzayas** [comment]
I have a working version of the new MPEB currently in the feature/NewEventBuilder branch. KTMultiPeakEventBuilder has been reduced to only the second piece, with its track slot replaced by a slot to accept multi-peak tracks which are produced by a new processor KTMultiPeakTrackBuilder. The event builder also re-emits signals containing each of the MPTs it received, with the event sequences assigned; but as the code is organized currently, these must be re-constructed from the individual tracks at the end. I have tried to verify that this is done correctly by comparing the output of all MPTs with multiplicity 2 and event sequence ID = 0 between the old and new event builders; currently I am experiencing about an 8% loss that I an unable to explain.

**Comment by evzayas** [comment]
Update: I found the bug which was causing an 8% discrepancy, and I have now verified that the new MPEB performs exactly as the old one, with the addition of individual multi-peak tracks for analysis.

**Comment by evzayas** [comment]
This change has been merged into develop so I am closing the issue.
